File: CleverFerretV2/core/ui/src/main/java/com/cleverferret/v2/core/ui/theme/KthemeThemeAdapterV1.java
package com.cleverferret.v2.core.ui.theme;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Adapter boundary between core:ui semantic roles and GitHub.com/kaleaon/ktheme themes.
 *
 * <p>Compatibility rules for future ktheme updates:
 * - Only depend on required baseline color fields.
 * - Preserve unknown/future fields in {@code passthroughRoles}.
 * - Keep semantic role mapping stable for feature modules.</p>
 */
public final class KthemeThemeAdapterV1 {
    private KthemeThemeAdapterV1() {}

    public record KthemeSnapshot(
            String id,
            boolean darkMode,
            String primary,
            String onPrimary,
            String background,
            String onBackground,
            String surface,
            String onSurface,
            String outline,
            String error,
            String onError,
            Map<String, String> passthroughRoles
    ) {}

    public record SemanticTheme(
            String id,
            boolean dark,
            Map<String, String> semanticColors,
            Map<String, String> passthroughRoles
    ) {}

    public static SemanticTheme adapt(KthemeSnapshot snapshot) {
        Map<String, String> semantic = new LinkedHashMap<>();
        semantic.put("color.action.primary", snapshot.primary());
        semantic.put("color.content.inverse", snapshot.onPrimary());
        semantic.put("color.surface.background", snapshot.background());
        semantic.put("color.content.primary", snapshot.onBackground());
        semantic.put("color.surface.elevated", snapshot.surface());
        semantic.put("color.content.secondary", snapshot.onSurface());
        semantic.put("color.border.default", snapshot.outline());
        semantic.put("color.state.error", snapshot.error());
        semantic.put("color.content.error", snapshot.onError());

        return new SemanticTheme(
                snapshot.id(),
                snapshot.darkMode(),
                Map.copyOf(semantic),
                snapshot.passthroughRoles() == null ? Map.of() : Map.copyOf(snapshot.passthroughRoles())
        );
    }
}

File: CleverFerretV2/core/ui/src/main/java/com/cleverferret/v2/core/ui/theme/KthemeThemeAdapterV1.kt
package com.cleverferret.v2.core.ui.theme

import com.cleverferret.v2.core.ui.tokens.DesignTokensV1

object KthemeThemeAdapterV1 {
    data class KthemeSnapshot(
        val id: String,
        val darkMode: Boolean,
        val primary: String,
        val onPrimary: String,
        val background: String,
        val onBackground: String,
        val surface: String,
        val onSurface: String,
        val outline: String,
        val error: String,
        val onError: String,
        val passthroughRoles: Map<String, String>,
    )

    enum class SemanticRole(val token: String) {
        ACTION_PRIMARY(DesignTokensV1.ColorRoles.ACTION_PRIMARY),
        CONTENT_INVERSE(DesignTokensV1.ColorRoles.CONTENT_INVERSE),
        SURFACE_BACKGROUND(DesignTokensV1.ColorRoles.SURFACE_BACKGROUND),
        CONTENT_PRIMARY(DesignTokensV1.ColorRoles.CONTENT_PRIMARY),
        SURFACE_ELEVATED(DesignTokensV1.ColorRoles.SURFACE_ELEVATED),
        CONTENT_SECONDARY(DesignTokensV1.ColorRoles.CONTENT_SECONDARY),
        BORDER_DEFAULT(DesignTokensV1.ColorRoles.BORDER_DEFAULT),
        STATE_ERROR(DesignTokensV1.ColorRoles.STATE_ERROR),
        CONTENT_ERROR(DesignTokensV1.ColorRoles.CONTENT_ERROR),
    }

    data class SemanticTheme(
        val id: String,
        val dark: Boolean,
        val semanticColors: Map<String, String>,
        val passthroughRoles: Map<String, String>,
    ) {
        fun color(role: SemanticRole): String = requireNotNull(semanticColors[role.token]) {
            "Missing semantic role ${role.token} in theme '$id'"
        }
    }

    /**
     * Feature UI modules consume only SemanticTheme, never KthemeSnapshot palette fields.
     */
    fun adapt(snapshot: KthemeSnapshot): SemanticTheme {
        val semantic = linkedMapOf(
            DesignTokensV1.ColorRoles.ACTION_PRIMARY to snapshot.primary,
            DesignTokensV1.ColorRoles.CONTENT_INVERSE to snapshot.onPrimary,
            DesignTokensV1.ColorRoles.SURFACE_BACKGROUND to snapshot.background,
            DesignTokensV1.ColorRoles.CONTENT_PRIMARY to snapshot.onBackground,
            DesignTokensV1.ColorRoles.SURFACE_ELEVATED to snapshot.surface,
            DesignTokensV1.ColorRoles.CONTENT_SECONDARY to snapshot.onSurface,
            DesignTokensV1.ColorRoles.BORDER_DEFAULT to snapshot.outline,
            DesignTokensV1.ColorRoles.STATE_ERROR to snapshot.error,
            DesignTokensV1.ColorRoles.CONTENT_ERROR to snapshot.onError,
        )
        return SemanticTheme(snapshot.id, snapshot.darkMode, semantic.toMap(), snapshot.passthroughRoles.toMap())
    }
}
